### PR TITLE
cubic: proportional rollback threshold

### PR DESCRIPTION
**Pull request by junhochoi**
Currently a fixed 10 packets threshold is used for detecting a small
loss. When cwnd is very high (more than 1000 packets, for example),
this makes the detection less sensitive. New approach is to make a
rollback threshold relative to cwnd (currently 20%) to make it work
with various size of cwnd in real time.

### CODE REVIEW

File: src/recovery/cubic.rs [hunk 1]
```diff
@@ -61,9 +61,13 @@ const BETA_CUBIC: f64 = 0.7;
 
 const C: f64 = 0.4;
 
-/// The packet count threshold to restore to the prior state if the
-/// lost packet count since the last checkpoint is less than the threshold.
-const RESTORE_COUNT_THRESHOLD: usize = 10;
+/// The packet count threshold (%) in relative to cwnd, to roll back to
+/// the prior state if the lost packet count since the last checkpoint
+/// is less than the threshold.
```
Comment (LPardue) [inline src/recovery/cubic.rs:66]: I find this sentence hard to parse but have no suggestion on improvements.

File: src/recovery/cubic.rs [hunk 2]
```diff
@@ -210,12 +214,16 @@ fn on_packet_acked(
     // <https://tools.ietf.org/id/draft-ietf-tcpm-rfc8312bis-00.html#section-4.9>
     //
     // When the recovery episode ends with recovering
-    // a few packets (less than RESTORE_COUNT_THRESHOLD), it's considered
-    // as spurious and restore to the previous state.
+    // a few packets (less than cwnd / mss * ROLLBACK_THRESHOLD_PERCENT(%)), it's
+    // considered as spurious and restore to the previous state.
     if r.congestion_recovery_start_time.is_some() {
         let new_lost = r.lost_count - r.cubic_state.prior.lost_count;
+        let rollback_threshold = (r.congestion_window / r.max_datagram_size) *
```
Comment (LPardue) [inline src/recovery/cubic.rs:221]: Is `rollback_threshold` a hypothetical value based on maximally packed packets?

Comment (junhochoi) [reply src/recovery/cubic.rs:221]: Right. Initially I was thinking to use bytes_in_flight but sometimes it can be very low and didn't work well.

File: src/recovery/cubic.rs [hunk 3]
```diff
@@ -61,9 +61,13 @@ const BETA_CUBIC: f64 = 0.7;
 
 const C: f64 = 0.4;
 
-/// The packet count threshold to restore to the prior state if the
-/// lost packet count since the last checkpoint is less than the threshold.
-const RESTORE_COUNT_THRESHOLD: usize = 10;
+/// The packet count threshold (%) in relative to cwnd, to roll back to
```
Comment (ghedo) [inline src/recovery/cubic.rs:64]:
```suggestion
/// Threshold for rolling back state, as percentage of lost packets relative to cwnd.
```

File: src/recovery/cubic.rs [hunk 4]
```diff
@@ -61,9 +61,13 @@ const BETA_CUBIC: f64 = 0.7;
...
+/// the prior state if the lost packet count since the last checkpoint
+/// is less than the threshold.
+const ROLLBACK_THRESHOLD_PERCENT: usize = 20;
+
+/// Minimum count of rollback threshold (packets)
```
Comment (ghedo) [inline src/recovery/cubic.rs:69]:
```suggestion
/// Minimum threshold for rolling back state, in number of packets.
```

File: src/recovery/cubic.rs [hunk 5]
```diff
@@ -61,9 +61,13 @@ const BETA_CUBIC: f64 = 0.7;
 
 const C: f64 = 0.4;
 
-/// The packet count threshold to restore to the prior state if the
-/// lost packet count since the last checkpoint is less than the threshold.
-const RESTORE_COUNT_THRESHOLD: usize = 10;
+/// The packet count threshold (%) in relative to cwnd, to roll back to
+/// the prior state if the lost packet count since the last checkpoint
```
Comment (ghedo) [inline src/recovery/cubic.rs:65]:
```suggestion
```

File: src/recovery/cubic.rs (hunk 1 above)
Comment (ghedo) [inline src/recovery/cubic.rs:66]:
```suggestion
```

**Review by ghedo — DISMISSED** [review]
LGTM, added some comments.

File: src/recovery/cubic.rs (hunk 1 above)
Comment (ghedo) [reply src/recovery/cubic.rs:66]: I added a suggestion. I don't think we need to repeat the whole thing, since the mechanism is described in the code where the constant is used.

**Review by ghedo — APPROVED** [review]
